pipeline/run_pipeline.py

```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
+            model_base.model,
+            harmless_train,
+            model_base.tokenize_instructions_fn,
+            model_base.refusal_toks,
+        )
+        harmful_train = filter_examples(
+            harmful_train, harmful_train_scores, 0, lambda x, y: x > y
+        )
+        harmless_train = filter_examples(
+            harmless_train, harmless_train_scores, 0, lambda x, y: x < y
+        )
 
     if cfg.filter_val:
-        harmful_val_scores = get_refusal_scores(model_base.model, harmful_val, model_base.tokenize_instructions_fn, model_base.refusal_toks)
-        harmless_val_scores = get_refusal_scores(model_base.model, harmless_val, model_base.tokenize_instructions_fn, model_base.refusal_toks)
-        harmful_val = filter_examples(harmful_val, harmful_val_scores, 0, lambda x, y: x > y)
-        harmless_val = filter_examples(harmless_val, harmless_val_scores, 0, lambda x, y: x < y)
-    
+        harmful_val_scores = get_refusal_scores(
+            model_base.model,
+            harmful_val,
+            model_base.tokenize_instructions_fn,
+            model_base.refusal_toks,
+        )
+        harmless_val_scores = get_refusal_scores(
+            model_base.model,
+            harmless_val,
+            model_base.tokenize_instructions_fn,
+            model_base.refusal_toks,
+        )
+        harmful_val = filter_examples(
+            harmful_val, harmful_val_scores, 0, lambda x, y: x > y
+        )
+        harmless_val = filter_examples(
+            harmless_val, harmless_val_scores, 0, lambda x, y: x < y
+        )
+
     return harmful_train, harmless_train, harmful_val, harmless_val
 
-def generate_and_save_candidate_directions(cfg, model_base, harmful_train, harmless_train):
+
+def generate_and_save_candidate_directions(
+    cfg, model_base, harmful_train, harmless_train
+):
     """Generate and save candidate directions."""
-    if not os.path.exists(os.path.join(cfg.artifact_path(), 'generate_directions')):
-        os.makedirs(os.path.join(cfg.artifact_path(), 'generate_directions'))
+    if not os.path.exists(os.path.join(cfg.artifact_path(), "generate_directions")):
+        os.makedirs(os.path.join(cfg.artifact_path(), "generate_directions"))
 
     mean_diffs = generate_directions(
         model_base,
         harmful_train,
         harmless_train,
-        artifact_dir=os.path.join(cfg.artifact_path(), "generate_directions"))
+        artifact_dir=os.path.join(cfg.artifact_path(), "generate_directions"),
+    )
 
-    torch.save(mean_diffs, os.path.join(cfg.artifact_path(), 'generate_directions/mean_diffs.pt'))
+    torch.save(
+        mean_diffs,
+        os.path.join(cfg.artifact_path(), "generate_directions/mean_diffs.pt"),
+    )
 
     return mean_diffs
 
-def select_and_save_direction(cfg, model_base, harmful_val, harmless_val, candidate_directions):
+
+def normalize_vector(vector):
+    """Normalize a given vector to have unit length."""
+    norm = torch.norm(vector)
```

Comment:
```suggestion
    norm = torch.linalg.vector_norm(vector)
```

Use `torch.linalg.vector_norm` here. `torch.norm` is deprecated for good reasons (like confusing matrix norm with vector norm): https://pytorch.org/docs/stable/generated/torch.norm.html

My guess is most likely this is a major bug in the code that is not letting us reproduce the paper's results.

Also add an assert statement to confirm the shape of `vector` is as you expect it to be for this function. Maybe `assert len(vector.shape) == 1`?